### PR TITLE
Enable the use of GCP Stackdriver

### PR DESCRIPTION
**Pull request by jtama**
Todos:

- [ ] Documentation
- [ ] Testing with InMemoryExporter

### CODE REVIEW

File: docs/src/main/asciidoc/deploying-to-google-cloud.adoc [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+== Connecting to CloudSQL
+
+Google CloudSQL managed service allows 4 kinds of connection :
+
+. Using public IP
+. Using private IP
+. Using Cloud SQL Proxy
+. Using service account
+
+The first two don't need anymore details, simply connect to you database following the link:datasource.adoc[datasource guide].
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-google-cloud.adoc]:
```suggestion
The first two don't need anymore details, simply connect to your database following the link:datasource[datasource guide].
```

File: docs/src/main/asciidoc/deploying-to-google-cloud.adoc [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
+////
+= Quarkus - Deploying on Google cloud platform
+
+include::./attributes.adoc[]
+
+This guide covers:
+
+* Connecting to a CloudSQL postgresql instance with quarkus-agroal and service account using CloudRun
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-google-cloud.adoc]:
```suggestion
* Connecting to a CloudSQL PostgreSQL instance with quarkus-agroal and service account using CloudRun
```
Maybe let's not use bullets for only one item?

File: docs/src/main/asciidoc/deploying-to-google-cloud.adoc [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+== Prerequisites
+
+For this guide you need:
+
+* roughly 1 hour
+* having access to an Google cloud platform project with owner rights.
+
+This guide will take as input an application developed in the link:datasource.adoc[datasource guide].
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-google-cloud.adoc]:
```suggestion
This guide will take as input an application developed in the link:datasource[datasource guide].
```

File: docs/src/main/asciidoc/deploying-to-google-cloud.adoc [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+Google CloudSQL managed service allows 4 kinds of connection :
+
+. Using public IP
+. Using private IP
+. Using Cloud SQL Proxy
+. Using service account
+
+The first two don't need anymore details, simply connect to you database following the link:datasource.adoc[datasource guide].
+The third allows you to connect as if locally, but is not available for AppEngine or CloudRun where you only deploy one artefact.
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-google-cloud.adoc]:
```suggestion
The third allows you to connect as if locally, but is not available for AppEngine or CloudRun where you only deploy one artifact.
```

File: docs/src/main/asciidoc/deploying-to-google-cloud.adoc [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+Please check link:https://cloud.google.com/sql/docs/postgres/external-connection-methods?hl=en[Connection options for external applications].
+
+This guide will help you through the fourth possibility : connecting using service account.
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-google-cloud.adoc]:
```suggestion
This guide will help you through the fourth possibility : connecting using a service account.
```

File: docs/src/main/asciidoc/deploying-to-google-cloud.adoc [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+== Adding necessary dependencies to your application.
+
+You will need to add the _Cloud SQL Postgres Socket Factory_ and postgresql driver dependencies.
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-google-cloud.adoc]:
```suggestion
You will need to add the _Cloud SQL Postgres Socket Factory_ and PostgreSQL driver dependencies.
```

File: docs/src/main/asciidoc/deploying-to-google-cloud.adoc [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+With maven :
+[source,xml, subs="attributes"]
+----
+<!-- https://mvnrepository.com/artifact/com.google.cloud.sql/postgres-socket-factory -->
+<dependency>
+    <groupId>com.google.cloud.sql</groupId>
+    <artifactId>postgres-socket-factory</artifactId>
+    <version>1.0.15</version>
+</dependency>
+<!-- https://mvnrepository.com/artifact/org.postgresql/postgresql -->
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-google-cloud.adoc]: Could we point to the more standard https://search.maven.org/ (or remove the links altogether)?

File: docs/src/main/asciidoc/deploying-to-google-cloud.adoc [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    <version>42.2.8</version>
+</dependency>
+----
+
+With Gradle :
+[source,groovy, subs="attributes"]
+----
+// https://mvnrepository.com/artifact/com.google.cloud.sql/postgres-socket-factory
+compile group: 'com.google.cloud.sql', name: 'postgres-socket-factory', version: '1.0.15'
+// https://mvnrepository.com/artifact/org.postgresql/postgresql
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-google-cloud.adoc]: Same here.

File: docs/src/main/asciidoc/deploying-to-google-cloud.adoc [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+compile group: 'org.postgresql', name: 'postgresql', version: '42.2.8'
+----
+
+== Configuring the CloudSQL instance
+
+If you haven't already please follow the link::https://cloud.google.com/sql/docs/postgres/create-instance[Creating instances guide].
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-google-cloud.adoc]:
```suggestion
If you haven't already, please follow the https://cloud.google.com/sql/docs/postgres/create-instance[Creating instances guide].
```

File: docs/src/main/asciidoc/deploying-to-google-cloud.adoc [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+[source,properties]
+--
+quarkus.datasource.url=jdbc:postgresql:///${DB_NAME}?socketFactory=com.google.cloud.sql.postgres.SocketFactory&cloudSqlInstance=${CLOUD_SQL_INSTANCE}
+quarkus.datasource.driver=org.postgresql.Driver
+quarkus.datasource.username = ${DB_USER}
+quarkus.datasource.password = ${DB_PASSWORD}
+--
+
+Prefere the use of environment variables that will allow you to connect to all your env depending on your runtime values
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-google-cloud.adoc]:
```suggestion
Prefere the use of environment variables that will allow you to connect to all your environment depending on your runtime values.
```

File: docs/src/main/asciidoc/deploying-to-google-cloud.adoc [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+[source,docker]
+--
+FROM gcr.io/distroless/java:11
+COPY target/*.jar /app/application-runner.jar
+COPY <your_service_account_key>.json /app
+WORKDIR /app
+CMD ["application-runner.jar","-Dquarkus.http.host=0.0.0.0", "-Dquarkus.http.port=$PORT"]
+--
+
+NOTE: `$PORT` value will be injected automatically by CloudRun and is to be used for HealthCheck
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-google-cloud.adoc]:
```suggestion
NOTE: `$PORT` value will be injected automatically by CloudRun and is to be used for HealthCheck.
```

File: docs/src/main/asciidoc/deploying-to-google-cloud.adoc [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+== Configuring the application
+
+Configure your quarkus application as follows :
+
+[source,properties]
+--
+quarkus.datasource.url=jdbc:postgresql:///${DB_NAME}?socketFactory=com.google.cloud.sql.postgres.SocketFactory&cloudSqlInstance=${CLOUD_SQL_INSTANCE}
+quarkus.datasource.driver=org.postgresql.Driver
+quarkus.datasource.username = ${DB_USER}
+quarkus.datasource.password = ${DB_PASSWORD}
```
Comment (gsmet) [inline docs/src/main/asciidoc/deploying-to-google-cloud.adoc]: It looks like you don't use the configuration properties you introduced in the extension? It is expected?

File: extensions/agroal/deployment/src/test/java/io/quarkus/agroal/test/GcpDataSourceConfigTest.java [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+import io.agroal.api.configuration.AgroalConnectionPoolConfiguration;
+import io.agroal.narayana.NarayanaTransactionIntegration;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class GcpDataSourceConfigTest {
+
+    //tag::injection[]
+    @Inject
+    AgroalDataSource defaultDataSource;
+    //end::injection[]
```
Comment (gsmet) [inline extensions/agroal/deployment/src/test/java/io/quarkus/agroal/test/GcpDataSourceConfigTest.java]: I don't think you need the tag here.

File: extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourceRuntimeConfig.java [hunk 14]
```diff
@@ -130,4 +130,16 @@
...
+     * When enabled jdbc url will be forged to enable google socket factory
+     */
+    @ConfigItem(defaultValue = "false")
+    public boolean useGcp;
+
+    /**
+     * Query executed to validate a connection.
+     */
+    @ConfigItem
+    public Optional<String> cloudSqlInstance;
```
Comment (gsmet) [inline extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourceRuntimeConfig.java]: I'm not entirely sure we will want to include those very specific properties in a generic extension. We need to discuss this.

File: extensions/arc/runtime/pom.xml [hunk 15]
```diff
@@ -1,52 +1,52 @@
...
+                            <groupId>io.quarkus</groupId>
+                            <artifactId>quarkus-extension-processor</artifactId>
+                            <version>${project.version}</version>
+                        </path>
+                    </annotationProcessorPaths>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
 </project>
```
Comment (gsmet) [inline extensions/arc/runtime/pom.xml]: Please don't reformat files.

File: extensions/arc/runtime/src/main/resources/META-INF/quarkus-extension.yaml [hunk 16]
```diff
@@ -3,12 +3,12 @@ name: "ArC"
...
+    - "arc"
+    - "cdi"
+    - "dependency-injection"
+    - "di"
   guide: "https://quarkus.io/guides/cdi-reference"
   categories:
-  - "core"
+    - "core"
   status: "stable"
   unlisted: true
```
Comment (gsmet) [inline extensions/arc/runtime/src/main/resources/META-INF/quarkus-extension.yaml]: Same here.

File: extensions/stackdriver/deployment/pom.xml [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-stackdriver-parent</artifactId>
+        <version>999-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>quarkus-stackdriver-deployment</artifactId>
+    <name>Opencensus - Stackdriver configurer - Deployment</name>
```
Comment (gsmet) [inline extensions/stackdriver/deployment/pom.xml]: Please follow the conventions for the name of the extension.

File: extensions/stackdriver/deployment/pom.xml [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+            <artifactId>quarkus-resteasy-deployment</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-rest-client-deployment</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-stackdriver</artifactId>
+            <version>${project.version}</version>
```
Comment (gsmet) [inline extensions/stackdriver/deployment/pom.xml]: The version should be in the BOM and not here.

File: extensions/stackdriver/deployment/pom.xml [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <annotationProcessorPaths>
+                        <path>
+                            <groupId>io.quarkus</groupId>
+                            <artifactId>quarkus-extension-processor</artifactId>
+                            <version>1.1.0.Final</version>
```
Comment (gsmet) [inline extensions/stackdriver/deployment/pom.xml]: This needs to be a variable.

File: extensions/stackdriver/pom.xml [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-extensions-parent</artifactId>
+        <version>999-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>quarkus-stackdriver-parent</artifactId>
+    <name>Opencensus - Stackdriver configurer - Parent</name>
```
Comment (gsmet) [inline extensions/stackdriver/pom.xml]: Same here about the conventions.

File: extensions/stackdriver/runtime/pom.xml [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-stackdriver-parent</artifactId>
+        <version>999-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>quarkus-stackdriver</artifactId>
+    <name>Opencensus - Stackdriver configurer - Runtime</name>
```
Comment (gsmet) [inline extensions/stackdriver/runtime/pom.xml]: Conventions.

File: extensions/stackdriver/runtime/pom.xml [hunk 22]
```diff
@@ -0,0 +1,219 @@
...
+    <properties>
+        <opencensus.version>0.24.0</opencensus.version>
```
Comment (gsmet) [inline extensions/stackdriver/runtime/pom.xml]: This should be in the BOM.

File: extensions/stackdriver/runtime/pom.xml [hunk 23]
```diff
@@ -0,0 +1,219 @@
...
+            <artifactId>quarkus-resteasy</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-rest-client</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.opentracing</groupId>
+            <artifactId>opentracing-api</artifactId>
+            <version>0.33.0</version>
```
Comment (gsmet) [inline extensions/stackdriver/runtime/pom.xml]: Same here.

File: extensions/stackdriver/runtime/pom.xml [hunk 24]
```diff
@@ -0,0 +1,219 @@
...
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-arc</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>jakarta.servlet</groupId>
+            <artifactId>jakarta.servlet-api</artifactId>
+        </dependency>
+
+        <!-- On démerde le caca de google -->
```
Comment (gsmet) [inline extensions/stackdriver/runtime/pom.xml]: Hum, I don't think we want that sort of comments.

File: extensions/stackdriver/runtime/pom.xml [hunk 25]
```diff
@@ -0,0 +1,219 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.google.j2objc</groupId>
+            <artifactId>j2objc-annotations</artifactId>
+            <version>1.3</version>
+        </dependency>
+        <dependency>
+            <groupId>com.google.auth</groupId>
+            <artifactId>google-auth-library-oauth2-http</artifactId>
+            <version>0.16.2</version>
```
Comment (gsmet) [inline extensions/stackdriver/runtime/pom.xml]: All the versions should be in the BOM

File: extensions/stackdriver/runtime/pom.xml [hunk 26]
```diff
@@ -0,0 +1,219 @@
...
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <annotationProcessorPaths>
+                        <path>
+                            <groupId>io.quarkus</groupId>
+                            <artifactId>quarkus-extension-processor</artifactId>
+                            <version>1.1.0.Final</version>
```
Comment (gsmet) [inline extensions/stackdriver/runtime/pom.xml]: Same here about the version.

File: extensions/stackdriver/runtime/src/main/java/io/quarkus/stackdriver/filter/SpanWrapper.java [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import io.opencensus.common.Scope;
+import io.opencensus.trace.Span;
+
+/**
+ * Wrapper class used for exchanging span between filters.
+ *
+ * @author Pavol Loffay
```
Comment (gsmet) [inline extensions/stackdriver/runtime/src/main/java/io/quarkus/stackdriver/filter/SpanWrapper.java]: We don't do `@author` tags.

File: extensions/stackdriver/runtime/src/main/java/io/quarkus/stackdriver/runtime/StackdriverProducer.java [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+import java.net.URI;
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.enterprise.context.Dependent;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.Produces;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Comment (gsmet) [inline extensions/stackdriver/runtime/src/main/java/io/quarkus/stackdriver/runtime/StackdriverProducer.java]: This should be JBoss Logging.

File: extensions/stackdriver/runtime/src/main/java/io/quarkus/stackdriver/runtime/configuration/StackdriverConfiguration.java [hunk 29]
```diff
@@ -0,0 +1,21 @@
...
+import io.quarkus.runtime.annotations.ConfigPhase;
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot(name = "gcp.stackdriver", phase = ConfigPhase.RUN_TIME)
+public class StackdriverConfiguration {
+
+    /**
+     * The GCP project ID
+     */
+    @ConfigItem(name = "project-id", defaultValue = "!")
```
Comment (gsmet) [inline extensions/stackdriver/runtime/src/main/java/io/quarkus/stackdriver/runtime/configuration/StackdriverConfiguration.java]: You don't need a name here.

File: extensions/stackdriver/runtime/src/main/java/io/quarkus/stackdriver/runtime/configuration/StackdriverConfiguration.java [hunk 30]
```diff
@@ -0,0 +1,21 @@
...
+    public String projectId;
+
+    /**
+     * The GCP project ID
```
Comment (gsmet) [inline extensions/stackdriver/runtime/src/main/java/io/quarkus/stackdriver/runtime/configuration/StackdriverConfiguration.java]: Documentation is incorrect.

File: extensions/stackdriver/runtime/src/main/java/io/quarkus/stackdriver/runtime/configuration/StackdriverConfiguration.java [hunk 31]
```diff
@@ -0,0 +1,21 @@
...
+     */
+    @ConfigItem(name = "credentials")
```
Comment (gsmet) [inline extensions/stackdriver/runtime/src/main/java/io/quarkus/stackdriver/runtime/configuration/StackdriverConfiguration.java]: No need for the name attribute here.

File: extensions/stackdriver/runtime/src/main/resources/META-INF/quarkus-gcp-stackdriver.yml [hunk 32]
```diff
@@ -0,0 +1,11 @@
...
+metadata:
+  keywords:
+    - "gcp"
+    - "opencensus"
+    - "stackdriver"
+    - "observability"
+  guide: "https://quarkus.io/guides/stackdrvier"
+  categories:
+    - "observability"
+  status: "draft"
```
Comment (gsmet) [inline extensions/stackdriver/runtime/src/main/resources/META-INF/quarkus-gcp-stackdriver.yml]: it should be experimental.

File: extensions/stackdriver/runtime/src/main/resources/META-INF/quarkus-gcp-stackdriver.yml [hunk 33]
```diff
@@ -0,0 +1,11 @@
+name: "GCP - Stackdriver extension"
+metadata:
+  keywords:
+    - "gcp"
+    - "opencensus"
+    - "stackdriver"
+    - "observability"
+  guide: "https://quarkus.io/guides/stackdrvier"
```
Comment (gsmet) [inline extensions/stackdriver/runtime/src/main/resources/META-INF/quarkus-gcp-stackdriver.yml]: Typo in `stackdrvier`

File: extensions/stackdriver/runtime/src/main/resources/META-INF/quarkus-gcp-stackdriver.yml [hunk 34]
```diff
@@ -0,0 +1,11 @@
+name: "GCP - Stackdriver extension"
```
Comment (gsmet) [inline extensions/stackdriver/runtime/src/main/resources/META-INF/quarkus-gcp-stackdriver.yml]:
```suggestion
name: "GCP - Stackdriver"
```

**Review by gsmet — CHANGES_REQUESTED** [review]
Thanks. I added a bunch of comments.

What's really missing is the big picture and what you are trying to achieve with all that. I have no idea what I reviewed :).

**Comment by jtama-op** [comment]
> Thanks. I added a bunch of comments.
> 
> What's really missing is the big picture and what you are trying to achieve with all that. I have no idea what I reviewed :).

@gsmet: Yes I agree, I need to add a proper documentation.
The extension is mimicking the opencensus extension to automate the use of opentrace with stackdriver exporter.

That said Opentrace and Opensencus are beeing merged into OpenTelemetry, and I am thinking to bump all dependencies.

In the meantime, quarkus is clearly missing extension to work with GCP which is sad considaring the cloud native nature of quarkus.

**Comment by gsmet** [comment]
Closing as out of date. Open a new PR if you still need that.

**Pull request closed without merging**